Topics in this notebook:

* loading data from a local `.csv` file
* what is *tidy* data?
* how to *tidy* untidy data

In this unit we'll start by loading local (i.e. the computer you are sitting in front of) data into Python.

***All of the following needs to be modified once we figure out our instructions to students***

* Download the file `wide_data.csv` from Canvas
* Put the file in a local folder.  
    * on your laptop you might choose either an OneDrive desktop app folder, or
    * a DropBox desktop app folder
    * on a lab computer you can choose your OneDrive folder - however when you move to another device this pathname won't work.



Load pandas:

In [2]:
import pandas as pd

To load the file you will need to know its pathname.

To copy a file pathname on a Mac:

* locate the file in the finder
* control-click on the file name (right click) and do not release
* press the option key
* choose copy "...." as Pathname
* you can now paste the pathname in a `read_csv` command

To copy a filepath name on a PC:

* navigate to the folder that contains the file
* press and hold the shift key
* right click on the file
* choose `copy as path`
* you can now paste the pathname in a `read_csv` command


Now, you're ready to upload the file with `read_csv`:

In [7]:
wide_data=pd.read_csv("/Users/janiskoscielniak/Dropbox/DS/wide_data.csv")
wide_data

Unnamed: 0  Unnamed: 0.1                                 category  \
0             0             0                     Aboriginal languages   
1             1             1  Non-Official & Non-Aboriginal languages   
2             2             2  Non-Official & Non-Aboriginal languages   
3             3             3  Non-Official & Non-Aboriginal languages   
4             4             4  Non-Official & Non-Aboriginal languages   
..          ...           ...                                      ...   
209         209           209  Non-Official & Non-Aboriginal languages   
210         210           210                     Aboriginal languages   
211         211           211  Non-Official & Non-Aboriginal languages   
212         212           212  Non-Official & Non-Aboriginal languages   
213         213           213  Non-Official & Non-Aboriginal languages   

                           language  Toronto  Montréal  Vancouver  Calgary  \
0      Aboriginal languages, n.o.s.       80        30         70       20   
1                         Afrikaans      985        90       1435      960   
2    Afro-Asiatic languages, n.i.e.      360       240         45       45   
3                        Akan (Twi)     8485      1015        400      705   
4                          Albanian    13260      2450       1090     1365   
..                              ...      ...       ...        ...      ...   
209                           Wolof      165      2440         30      120   
210                      Woods Cree        5         0         20       10   
211               Wu (Shanghainese)     5290      1025       4330      380   
212                         Yiddish     3355      8960        220       80   
213                          Yoruba     3380       210        190     1430   

     Edmonton  
0          25  
1         575  
2          65  
3         885  
4         770  
..        ...  
209       130  
210       155  
211       235  
212        55  
213       700  

[214 rows x 9 columns]

We always want data to be in a *tidy* format.  (The term *tidy* is borrowed from R.) *Tidy* data is in a format where:

* each row is an observation
* each column is a variable
* each cell is a value

The data we just uploaded is *un-tidy*!  Why?  The names of cities, Toronto, Montréal, etc, are not *variables*. Rather, they are values for a variable we'll call `City`.

So, for example, the information contained in the first row of the table `wide_data` should be 5 rows. There are 5 *observations*: one for each *value* (city name) of the *variable* `City`. 

We'll use the panda command `melt` to make the data *tidy*.  It's important to be able to figure out how to read and use python/panda documentation.  We'll practice with this example.  
* Open the website: https://pandas.pydata.org/docs/index.html
* Type `melt` in the search box on the left.
* Choose `pandas.melt`


It takes some practice to be able to make sense of this sort of documentation:

* `frame` is the name of the data frame we want to *melt*: for this example it's `wide_date`
* `id_vars` are the columns we want to leave alone because they are varables: for this example they are `category` and `language`
* `value_vars` are the titles on columns that **are not** variables: for this example these are the names of the cities
* Notice there is one column that isn't doing anything: `Unnamed:` We'll just ignore it for now and hope it goes away.  If it doesn't we'll make it go away later.
* `var_name` is the name of the varable we'd like to create whose values are the names of cities
* `value_name` is the title we'll give the variable whose value is the number of speakers

Below is the command with all arguments completed.  Take some time to make sense of the command.  Remember: running a command you don't understand will not help you in the long run.

In [8]:
tidy_data=pd.melt(wide_data, id_vars=["category", "language"], 
                  value_vars=["Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton"], 
                  var_name=["City"],
                  value_name="number_speakers")

In [9]:
tidy_data

category                        language  \
0                        Aboriginal languages    Aboriginal languages, n.o.s.   
1     Non-Official & Non-Aboriginal languages                       Afrikaans   
2     Non-Official & Non-Aboriginal languages  Afro-Asiatic languages, n.i.e.   
3     Non-Official & Non-Aboriginal languages                      Akan (Twi)   
4     Non-Official & Non-Aboriginal languages                        Albanian   
...                                       ...                             ...   
1065  Non-Official & Non-Aboriginal languages                           Wolof   
1066                     Aboriginal languages                      Woods Cree   
1067  Non-Official & Non-Aboriginal languages               Wu (Shanghainese)   
1068  Non-Official & Non-Aboriginal languages                         Yiddish   
1069  Non-Official & Non-Aboriginal languages                          Yoruba   

          City  number_speakers  
0      Toronto               80  
1      Toronto              985  
2      Toronto              360  
3      Toronto             8485  
4      Toronto            13260  
...        ...              ...  
1065  Edmonton              130  
1066  Edmonton              155  
1067  Edmonton              235  
1068  Edmonton               55  
1069  Edmonton              700  

[1070 rows x 4 columns]

In [35]:
tidy_data.to_csv("/Users/janiskoscielniak/Dropbox/DS/tidy_langs1.csv")

One quick and dirty way to make sure the command *might* have done the right thing is to consider the number of rows.  There were 214 rows in the original table.  Our tidy table should have $214\cdot5=1070$ rows. 

#### Your turn:

> Use the tidy data frame `tidy_data` to generate a table of Aboriginal languages spoken in Vancouver, sorted in descending order by number of speakers. (Use the python commands you learned in the previous notebook.)

In [10]:
tidy_data[(tidy_data['category']=='Aboriginal languages') & 
          (tidy_data['City']=='Vancouver') & 
          (tidy_data['number_speakers'] != 0)].sort_values('number_speakers', ascending=False)


category                      language       City  \
468  Aboriginal languages                  Cree, n.o.s.  Vancouver   
428  Aboriginal languages  Aboriginal languages, n.o.s.  Vancouver   
566  Aboriginal languages                       Ojibway  Vancouver   
495  Aboriginal languages             Gitxsan (Gitksan)  Vancouver   
503  Aboriginal languages                    Halkomelem  Vancouver   
585  Aboriginal languages      Salish languages, n.i.e.  Vancouver   
577  Aboriginal languages                   Plains Cree  Vancouver   
458  Aboriginal languages                       Carrier  Vancouver   
559  Aboriginal languages                       Nisga'a  Vancouver   
624  Aboriginal languages                     Tsimshian  Vancouver   
453  Aboriginal languages                     Blackfoot  Vancouver   
464  Aboriginal languages                     Chilcotin  Vancouver   
607  Aboriginal languages                      Squamish  Vancouver   
638  Aboriginal languages                    Woods Cree  Vancouver   
474  Aboriginal languages                        Dakota  Vancouver   
476  Aboriginal languages                          Dene  Vancouver   
619  Aboriginal languages        Thompson (Ntlakapamux)  Vancouver   
517  Aboriginal languages                     Inuktitut  Vancouver   
611  Aboriginal languages                   Swampy Cree  Vancouver   
533  Aboriginal languages          Kwakiutl (Kwak'wala)  Vancouver   
506  Aboriginal languages                      Heiltsuk  Vancouver   
500  Aboriginal languages                        Haisla  Vancouver   
564  Aboriginal languages       Nuu-chah-nulth (Nootka)  Vancouver   
593  Aboriginal languages       Shuswap (Secwepemctsin)  Vancouver   
498  Aboriginal languages                      Gwich'in  Vancouver   
499  Aboriginal languages                         Haida  Vancouver   
548  Aboriginal languages                        Michif  Vancouver   
536  Aboriginal languages                      Lillooet  Vancouver   
565  Aboriginal languages                      Oji-Cree  Vancouver   
551  Aboriginal languages                        Mohawk  Vancouver   
567  Aboriginal languages                      Okanagan  Vancouver   
547  Aboriginal languages                       Mi'kmaq  Vancouver   
586  Aboriginal languages                Sarsi (Sarcee)  Vancouver   
588  Aboriginal languages                        Sekani  Vancouver   
434  Aboriginal languages                     Algonquin  Vancouver   
597  Aboriginal languages      Siouan languages, n.i.e.  Vancouver   
598  Aboriginal languages                Slavey, n.o.s.  Vancouver   
445  Aboriginal languages          Babine (Wetsuwet'en)  Vancouver   
532  Aboriginal languages                       Kutenai  Vancouver   
516  Aboriginal languages       Inuit languages, n.i.e.  Vancouver   
440  Aboriginal languages  Athabaskan languages, n.i.e.  Vancouver   

     number_speakers  
468              145  
428               70  
566               60  
495               50  
503               40  
585               40  
577               40  
458               35  
559               30  
624               25  
453               25  
464               25  
607               20  
638               20  
474               20  
476               20  
619               15  
517               15  
611               15  
533               15  
506               15  
500               10  
564               10  
593               10  
498               10  
499               10  
548               10  
536               10  
565                5  
551                5  
567                5  
547                5  
586                5  
588                5  
434                5  
597                5  
598                5  
445                5  
532                5  
516                5  
440                5

#### Common sense check/warning:

> There is something suspicious about the number of speakers.  What is it?

#### Answer:

> All the values are multiples of 5. This seems unlikely.  It may not matter in further analysis, or it might.  It's always important to look at and think carefully about results you generate.

#### Your turn:

> Download the file `long_data.csv` from Canvas and put it in your IntroToDS folder.

In [11]:
long_data=pd.read_csv("/Users/janiskoscielniak/Dropbox/DS/long_data.csv")
long_data

Unnamed: 0     region                                 category  \
0              0   Montréal                     Aboriginal languages   
1              1   Montréal                     Aboriginal languages   
2              2    Toronto                     Aboriginal languages   
3              3    Toronto                     Aboriginal languages   
4              4    Calgary                     Aboriginal languages   
...          ...        ...                                      ...   
2135        2135    Calgary  Non-Official & Non-Aboriginal languages   
2136        2136   Edmonton  Non-Official & Non-Aboriginal languages   
2137        2137   Edmonton  Non-Official & Non-Aboriginal languages   
2138        2138  Vancouver  Non-Official & Non-Aboriginal languages   
2139        2139  Vancouver  Non-Official & Non-Aboriginal languages   

                          language          type  count  
0     Aboriginal languages, n.o.s.  most_at_home     15  
1     Aboriginal languages, n.o.s.  most_at_work      0  
2     Aboriginal languages, n.o.s.  most_at_home     50  
3     Aboriginal languages, n.o.s.  most_at_work      0  
4     Aboriginal languages, n.o.s.  most_at_home      5  
...                            ...           ...    ...  
2135                        Yoruba  most_at_work      0  
2136                        Yoruba  most_at_home    280  
2137                        Yoruba  most_at_work      0  
2138                        Yoruba  most_at_home     40  
2139                        Yoruba  most_at_work      0  

[2140 rows x 6 columns]

#### Question:

> This table is also untidy, but of a different reason. Why is this table untidy? Hint: each pair of rows should be one row.

#### Answer:

> `most_at_home` and `most_at_work` are variables and should be the titles for columns.  The values of these variables are shown next to the titles on each row.  So, for example, the first two rows should be a single row.

To tidy this data frame we need to combine two rows into a single row.  We'll use `pivot` to do this. The variables are: `region`, `category`, `language`, `most_at_home`, and `most_at_work`.

So, we are going to group rows together by:
* `region`,
* `category`, and 
* `language`

and convert the column `type` into two new columns: `most_at_home` and `most_at_work` whose values come from the column labeled `count`.

The code below *almost* does what we want:

In [17]:
tidy_data2=pd.pivot(long_data, 
                    index=['region', 'category', 'language'],
                    columns=['type'], 
                    values=['count'])
tidy_data2

count  \
type                                                                           most_at_home   
region    category                                language                                    
Calgary   Aboriginal languages                    Aboriginal languages, n.o.s.            5   
                                                  Algonquian languages, n.i.e.            0   
                                                  Algonquin                               0   
                                                  Athabaskan languages, n.i.e.            0   
                                                  Atikamekw                               0   
...                                                                                     ...   
Vancouver Non-Official & Non-Aboriginal languages Wu (Shanghainese)                    2495   
                                                  Yiddish                                10   
                                                  Yoruba                                 40   
          Official languages                      English                           1622735   
                                                  French                               8630   

                                                                                             
type                                                                           most_at_work  
region    category                                language                                   
Calgary   Aboriginal languages                    Aboriginal languages, n.o.s.            0  
                                                  Algonquian languages, n.i.e.            0  
                                                  Algonquin                               0  
                                                  Athabaskan languages, n.i.e.            0  
                                                  Atikamekw                               0  
...                                                                                     ...  
Vancouver Non-Official & Non-Aboriginal languages Wu (Shanghainese)                      45  
                                                  Yiddish                                 0  
                                                  Yoruba                                  0  
          Official languages                      English                           1330555  
                                                  French                               3245  

[1070 rows x 2 columns]

This is a nicely formated table that show the information we want. **But**, it isn't a data frame.  Notice there are only 2 columns/variables!  We should have 5.

The command `reset_index()` takes care of this issue.

In [18]:
tidy_data3=tidy_data2.reset_index()
tidy_data3

region                                 category  \
type                                                       
0       Calgary                     Aboriginal languages   
1       Calgary                     Aboriginal languages   
2       Calgary                     Aboriginal languages   
3       Calgary                     Aboriginal languages   
4       Calgary                     Aboriginal languages   
...         ...                                      ...   
1065  Vancouver  Non-Official & Non-Aboriginal languages   
1066  Vancouver  Non-Official & Non-Aboriginal languages   
1067  Vancouver  Non-Official & Non-Aboriginal languages   
1068  Vancouver                       Official languages   
1069  Vancouver                       Official languages   

                          language        count               
type                               most_at_home most_at_work  
0     Aboriginal languages, n.o.s.            5            0  
1     Algonquian languages, n.i.e.            0            0  
2                        Algonquin            0            0  
3     Athabaskan languages, n.i.e.            0            0  
4                        Atikamekw            0            0  
...                            ...          ...          ...  
1065             Wu (Shanghainese)         2495           45  
1066                       Yiddish           10            0  
1067                        Yoruba           40            0  
1068                       English      1622735      1330555  
1069                        French         8630         3245  

[1070 rows x 5 columns]

Notice: each time we modified the data we gave the result a new name.  Whether you do this, or use the same name each time depends on the situation.  With new names we can access all the intermediate results if we need to. On the other hand, it can get confusing when there are names that aren't needed or used.

Unfortunately, there is still one annoying *problem*.  The names of the columns/variables are each a list of 2 names!  Execute the following command to see the names of the columns.  Notice each column has 2 names! 

In [19]:
list(tidy_data3.columns)

[('region', ''),
 ('category', ''),
 ('language', ''),
 ('count', 'most_at_home'),
 ('count', 'most_at_work')]

This situation might be useful in some contexts, but we'll simplify the column names:

In [20]:
tidy_data3.columns=['region', 'category', 'language', 'most_at_home', 'most_at_work']
tidy_data3

region                                 category  \
0       Calgary                     Aboriginal languages   
1       Calgary                     Aboriginal languages   
2       Calgary                     Aboriginal languages   
3       Calgary                     Aboriginal languages   
4       Calgary                     Aboriginal languages   
...         ...                                      ...   
1065  Vancouver  Non-Official & Non-Aboriginal languages   
1066  Vancouver  Non-Official & Non-Aboriginal languages   
1067  Vancouver  Non-Official & Non-Aboriginal languages   
1068  Vancouver                       Official languages   
1069  Vancouver                       Official languages   

                          language  most_at_home  most_at_work  
0     Aboriginal languages, n.o.s.             5             0  
1     Algonquian languages, n.i.e.             0             0  
2                        Algonquin             0             0  
3     Athabaskan languages, n.i.e.             0             0  
4                        Atikamekw             0             0  
...                            ...           ...           ...  
1065             Wu (Shanghainese)          2495            45  
1066                       Yiddish            10             0  
1067                        Yoruba            40             0  
1068                       English       1622735       1330555  
1069                        French          8630          3245  

[1070 rows x 5 columns]

In [36]:
tidy_data3.to_csv("/Users/janiskoscielniak/Dropbox/DS/tidy_langs2.csv")

#### Your turn:

> Use tidy_data3 to generate a list of Aboriginal languages spoken in Vancouver sorted in descending order by number of speakers who speak the language mostly at work.

In [21]:
tidy_data3[(tidy_data3['region']=='Vancouver') & 
           (tidy_data3['category']=='Aboriginal languages')].sort_values('most_at_work', ascending=False)

region              category                      language  \
868  Vancouver  Aboriginal languages                  Cree, n.o.s.   
856  Vancouver  Aboriginal languages  Aboriginal languages, n.o.s.   
899  Vancouver  Aboriginal languages                       Ojibway   
904  Vancouver  Aboriginal languages      Salish languages, n.i.e.   
903  Vancouver  Aboriginal languages                   Plains Cree   
..         ...                   ...                           ...   
884  Vancouver  Aboriginal languages          Kwakiutl (Kwak'wala)   
885  Vancouver  Aboriginal languages                      Lillooet   
886  Vancouver  Aboriginal languages                      Malecite   
887  Vancouver  Aboriginal languages                       Mi'kmaq   
922  Vancouver  Aboriginal languages                    Woods Cree   

     most_at_home  most_at_work  
868            15            10  
856            15             0  
899             5             0  
904             0             0  
903             0             0  
..            ...           ...  
884             5             0  
885             5             0  
886             0             0  
887             0             0  
922             5             0  

[67 rows x 5 columns]

#### Question:

> How many aboriginal languages are spoken at work in Vancouver?

#### Homework:

> Download the file `sea_temps.csv` from Canvas and put it in your DS folder. The data is of sea surface temperatures in Departure Bay, Canada.

> Upload the file into this notebook.

> This data frame is not tidy. Why?

> Tidy the data frame.

> Generate a list of all months/years when the temperature was 16 degrees Celsius or more.

In [23]:
sea_temps=pd.read_csv("/Users/janiskoscielniak/Dropbox/DS/sea_temps.csv")
sea_temps

Unnamed: 0  Year  Jan  Feb  Mar  Apr   May   Jun   Jul   Aug   Sep   Oct  \
0             0  1914  7.2  NaN  NaN  NaN   NaN   NaN   NaN   NaN  11.1  10.0   
1             1  1915  5.6  6.6  7.5  9.0   9.9  12.5  14.7  15.8  14.0   8.2   
2             2  1916  1.2  0.1  3.5  6.5   8.0  12.0  13.1  14.0  11.4   7.6   
3             3  1917  3.8  2.8  4.4  5.4   8.3  11.0  13.7  12.2  10.0   8.6   
4             4  1918  3.7  3.9  4.6  6.0   9.3  11.2  13.1  14.5  13.8   9.1   
..          ...   ...  ...  ...  ...  ...   ...   ...   ...   ...   ...   ...   
100         100  2014  4.4  3.1  3.7  7.7  10.1  12.0  15.9  16.8  11.7  10.6   
101         101  2015  6.3  8.0  8.0  8.9  11.0  15.5  13.8  13.4  11.6  11.3   
102         102  2016  6.0  7.1  8.4  9.8  13.0  14.2  14.6  14.6  12.6  10.8   
103         103  2017  5.6  4.8  7.1  7.9  10.5  12.4  15.3  15.3  13.1  10.2   
104         104  2018  6.2  6.0  7.1  8.2   NaN   NaN   NaN   NaN   NaN   NaN   

     Nov  Dec  
0    7.3  6.3  
1    4.4  4.1  
2    5.4  3.5  
3    7.0  4.9  
4    6.7  5.0  
..   ...  ...  
100  7.1  7.1  
101  8.1  6.8  
102  8.2  5.5  
103  8.8  6.9  
104  NaN  NaN  

[105 rows x 14 columns]

In [24]:
tidy_temps=pd.melt(sea_temps, id_vars=["Year"], 
                  value_vars=['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'], 
                  var_name=["Month"],
                  value_name="Temperature")
tidy_temps

Year Month  Temperature
0     1914   Jan          7.2
1     1915   Jan          5.6
2     1916   Jan          1.2
3     1917   Jan          3.8
4     1918   Jan          3.7
...    ...   ...          ...
1255  2014   Dec          7.1
1256  2015   Dec          6.8
1257  2016   Dec          5.5
1258  2017   Dec          6.9
1259  2018   Dec          NaN

[1260 rows x 3 columns]

In [37]:
tidy_temps.to_csv("/Users/janiskoscielniak/Dropbox/DS/tidy_temps.csv")

In [70]:
12*105

1260

In [25]:
tidy_temps[tidy_temps['Temperature']>=16]

Year Month  Temperature
641  1925   Jul         16.1
644  1928   Jul         16.1
653  1937   Jul         17.0
666  1950   Jul         16.4
674  1958   Jul         17.6
683  1967   Jul         16.5
691  1975   Jul         16.1
729  2013   Jul         17.0
744  1923   Aug         16.8
749  1928   Aug         16.4
752  1931   Aug         16.6
753  1932   Aug         16.4
756  1935   Aug         16.6
757  1936   Aug         16.1
763  1942   Aug         16.1
788  1967   Aug         17.0
793  1972   Aug         16.4
819  1998   Aug         16.9
824  2003   Aug         16.4
835  2014   Aug         16.8

#### Homework:

> Download the file `babynames.csv` from Canvas and put it in your DS folder. The file show the number of babies with a certain name 

> The file consists of all names of babies born in the USA from 1880 through 2017, together with the sex assigned to the baby, the number of babies of that sex given that name in that year, and the proportion of babies with that sex in that year that were given the name. Only names that were given to five or more babies are included.

> How many babies were given the name Carl in the year 1992?

> The answer to the question above illustrates why this data is not tidy.

> Tidy the data set. (You can ignore the variable `prop`.)

> Use the tidy data set to answer the question:  How many babies with your name and year of birth were assigned the gender of female at birth?  How many male?  (There are almost 2 million rows: use python filter commands to find the right row.)

In [26]:
babynames=pd.read_csv("/Users/janiskoscielniak/Dropbox/DS/babynames.csv")

In [27]:
babynames[(babynames['name']=='Carl')  & (babynames['year']==1992)]

year sex  name     n      prop
1127022  1992   F  Carl    11  0.000005
1135220  1992   M  Carl  1779  0.000848

In [28]:
babynames_wider=pd.pivot(babynames, 
                    index=['year', 'name'],
                    columns=['sex'], 
                    values=['n'])

In [29]:
df_names=babynames_wider.reset_index()

In [30]:
df_names.columns=['year', 'name', 'F', 'M']
df_names

year    name     F      M
0        1880   Aaron   NaN  102.0
1        1880      Ab   NaN    5.0
2        1880   Abbie  71.0    NaN
3        1880  Abbott   NaN    5.0
4        1880    Abby   6.0    NaN
...       ...     ...   ...    ...
1756279  2017  Zyriah  25.0    NaN
1756280  2017   Zyrie   NaN    5.0
1756281  2017   Zyron   NaN   14.0
1756282  2017   Zyrus   NaN    7.0
1756283  2017    Zyva   9.0    NaN

[1756284 rows x 4 columns]

In [31]:
df_names[(df_names['name']=='Carl') & (df_names['year'] ==1966)]

year  name     F       M
581938  1966  Carl  44.0  5226.0

In [38]:
df_names.to_csv("/Users/janiskoscielniak/Dropbox/DS/tidy_babies.csv")